README.md
# ecc-clj

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-10-22
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-10-22
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/ecc-clj/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/ecc-clj/compare/0.1.0...0.1.1


test/ecc_clj/qr_test.clj
(ns ecc-clj.qr-test
  (:require [ecc-clj.qr :refer :all]
            [clojure.test :refer :all]))

(deftest coord-sanity-check
  (testing "coordinates add up"
    (is
     (= 26 (/ (count v1-bit-order) 8)))
    (is
     (= 30 (count v1-fmt-order)))
    (is (= (* 21 21)
           (+ (count v1-fmt-order)
              (count fixed0s)
              (count fixed1s)
              (count v1-bit-order)))))
  (testing "coordinates don't intersect"
    (is (empty? (clojure.set/intersection
                 (set fixed0s) (set fixed1s))))
    (is (empty? (clojure.set/intersection
                 (set fixed0s) (set v1-data-order))))
    (is (empty? (clojure.set/intersection
                 (set fixed1s) (set v1-data-order))))
    ))

(deftest data-encoding-test
  (let [msg "www.wikipedia.org"
        bits (vec (encode-bytes (map int msg)))]
    (testing "length is 26 bytes"
      (is (= (* 26 8) (count bits))))
    (testing "encoding bits"
      (is (= [0 1 0 0]
             (subvec bits 0 4))))
    (testing "length bits"
      (is (= [0 0 0 1 0 0 0 1]
             (subvec bits 4 12))))
    (testing "end msg bits"
      (let [lastbyte (* 18 8)]
       (is (= [0 0 0 0]
              (subvec bits (+ 4 lastbyte) (+ 8 lastbyte))))))
    ))


test/ecc_clj/core_test.clj
(ns ecc-clj.core-test
  (:require [clojure.test :refer :all]
            [ecc-clj.core :refer :all]
            [ecc-clj.poly :as p]))

(deftest mod-inverses
  (testing "mod inverses"
    (let [p 17]
      (is
       (every? #(= 1 %)
               (map
                (fn [[a b]] (mod (* a b) p))
                (seq (p/mod-invs p))))))))

(deftest RS-7-5-test
  (testing "encoding"
    (let [data [5 2 3 1 6]]
      (is (= (encode data RS-7-5 GF8)
             [6 6 5 2 3 1 6]))))
  (testing "decoding without errors"
    (let [data [0 4 3 0 1]]
      (is (= data
             (-> data
                 (encode RS-7-5 GF8)
                 (RS-7-5-decode))))))
  (testing "decoding with errors"
    (let [encoded-poly [6 6 5 2 3 1 6]
          data-poly [5 2 3 1 6]]
      (is
       (every? #(= data-poly %)
        (for [err (single-errors 7 8)]
          (let [err-poly (p/+ encoded-poly err GF8)]
            (RS-7-5-decode err-poly)))))
      )))

(deftest RS-7-3-test
  (testing "encoding"
    (let [data [1 6 3]]
      (is (= (encode data RS-7-3 GF8)
             [0 2 2 4 1 6 3])))))


test/ecc_clj/poly_test.clj
(ns ecc-clj.poly-test
  (:require [ecc-clj.poly :as p]
            [ecc-clj.core :refer [GF2]]
            [clojure.test :refer :all]))

(deftest poly-mod-test
  (testing "integer polynomial remainder"
    (is (= (p/mod [-4 0 -2 1] [-3 1]) [5]))
    (is (= (p/mod [-3 2 -1 1] [-2 1]) [5]))
    (is (= (p/mod [-4 4 0 5 6] [-1 1 2]) [-3 4]))
    (is (= (p/mod [0 0 0 1] [1 0 1 1]) [-1 0 -1])))
  (testing "mod2 polynomial remainder"
    (let [f GF2]
      (is (= (p/mod [0 1 0 0 1] [1 1 0 1] f) [0 0 1]))
      )))

(deftest poly-quot-test
  (testing "divide binary polynomials"
    (let [F GF2
          p1 [0 1 0 0 1]
          p2 [1 1 0 1]
          prod (p/* p1 p2 F)
          [quot1 rem1] (p/quot-rem prod p2 GF2)
          [quot2 rem2] (p/quot-rem prod p1 GF2)]
      (is (and (= p1 quot1) (empty? rem1)))
      (is (and (= p2 quot2) (empty? rem2)))
      )))

(deftest binary-mod-test
  (testing "bit-polynomial remainder"
    (is (= 4
           (p/binmod
            (p/parse-bin "11010011101100000")
            (p/parse-bin "1011")))))

  (testing "binary polynomial mod functions agree"
    (let [p1 [0 1 0 0 1]
          p2 [1 1 0 1]
          p1-bits (p/vec-to-bits p1)
          p2-bits (p/vec-to-bits p2)]
      (is (= (p/mod p1 p2 GF2)
             (vec (reverse
                   (p/bits-to-vec
                    (p/binmod p1-bits p2-bits))))
             [0 0 1]))
      )))

(deftest char2-invs-test
  (testing "char2 field inverses"
    (let [prim 2
          poly (p/parse-bin "10011")
          invs (p/char2-invs prim poly)]
      (is (every? #(= 1 %)
        (map
         (fn [[a b]]
           (p/binmod (p/bin* a b) poly))
         (seq invs)))))
    ))

(deftest eval-test
  (testing "eval integer polynomial"
    (let [poly [3 0 2 1]]
      (is (= 3 (p/evaluate poly 0)))
      (is (= 6 (p/evaluate poly 1)))
      (is (= 19 (p/evaluate poly 2))))
    ))

(deftest poly-sub-test
  (testing "integer polynomial addition"
    (is (= (p/- [0 1] [1 -1 2]) [-1 2 -2]))
    (is (= (p/+ [0 4 -2 1] [0 -4 2 2]) [0 0 0 3]))))

(deftest poly-mul-test
  (testing "integer polynomial multiplication"
    (is (= (p/* [1 -1] [0 2 1]) [0 2 -1 -1]))
    (is (= (p/* [12 34 56] [0 1]) [0 12 34 56])))
  (testing "mod2 polynomial multiplication"
    (is (= (p/* [0 1 1] [1 1 1] GF2) [0 1 0 0 1]))))


src/ecc_clj/core.clj
(ns ecc-clj.core
  (:require [ecc-clj.poly :as p]
            [ecc-clj.linalg :as la]
            [taoensso.tufte :as tufte :refer [p profiled profile]]))

(def GF2
  {:unit 1
   :zero 0
   :+ bit-xor
   :- bit-xor
   :* bit-and
   :/ bit-and})

(def char2-primitive
  "The polynomial fields are constructed so that
  the polynomial x represented by 0b10 is primitive."
  2)

(def GF256
  "GF(255) constructed as GF2[x]/<x^8+x^7+x^2+x+1>"
  (let [GF2-poly (p/parse-bin "110000111")]
    (p/char2-field char2-primitive GF2-poly)))

(def GF16
  "GF(16) constructed as GF2[x]/<x^4+x+1>"
  (let [GF2-poly (p/parse-bin "10011")]
    (p/char2-field char2-primitive GF2-poly)))

(def GF8
  "GF(8) constructed as GF2[x]/<x^3+x+1>"
  (let [GF2-poly (p/parse-bin "1011")]
    (p/char2-field char2-primitive GF2-poly)))

(defn encode
  "systematic encoding: the data polynomial is the
  highest k coefficients and (n-k) lower are check symbols"
  [data-poly gen-poly & [field]]
  (let [field (or field p/default-field)
        shifted (p/shift-right
                 data-poly (dec (count gen-poly)))
        rem (p/mod shifted gen-poly field)]
    (p/+ shifted rem field)))

(defn single-errors
  "all single-error error polynomials"
  [len base]
  (for [i (range len)
        d (range 1 base)]
    (p/shift-right [d] i)))

(defn double-errors
  "all double-error error polynomials"
  [len base]
  (for [i (range len)
        j (range len)
        di (range 1 base)
        dj (range 1 base)
        :when (< i j)]
    (p/strip
     (assoc (vec (repeat len 0))
            i di
            j dj))))

(def RS-7-5
  "(7,5) Reed-Solomon code over GF(8) with
  generating polynomial
  g(x) = (x-w)(x-w^2), where w=2 "
  (p/* [2 1] [4 1] GF8))

(def RS-7-3
  "(7,3) Reed-Solomon code over GF(8)
  with generating polynomial
  g(x) = (x-w^4)(x-w^5)(x-w^6)(x-w^7), where w=2"
  (reduce
   (fn [p1 p2] (p/* p1 p2 GF8))
   [[7 1] [3 1] [6 1] [1 1]]))

;;; Small codes can be decoded with a lookup table
;;; of error-polynomial remainders
(def RS-7-5-table
  "Decoding table for (7,5) Reed-Solomon code.
  The polynomials are encoded as integers"
  (let [errors (single-errors 7 8)
        rems (map #(p/mod % RS-7-5 GF8) errors)
        to-int #(p/digits-to-int % 8)]
    (zipmap
     (map to-int rems)
     (map to-int errors))))

(def RS-7-3-table
  (let [errors (concat
                (single-errors 7 8)
                (double-errors 7 8))
        rems (map #(p/mod % RS-7-3 GF8) errors)
        to-int #(p/digits-to-int % 8)]
    (zipmap
     (map to-int rems)
     (map to-int errors))))

(defn RS-7-decode
  "decode a n=7 Reed Solomon code over GF8
  given the generating polynomial and decoding table"
  [gen-poly decoding-tbl data]
  (let [deg (dec (count gen-poly))
        extract-data (fn [v] (subvec v deg))

        rem (p/mod data gen-poly GF8)
        rem-int (p/digits-to-int rem 8)
        error (decoding-tbl rem-int)]
    (cond
      ;; no errors
      (empty? rem) (extract-data data)
      ;; uncorrectable error
      (nil? error) nil
      :else
      ;; correctable error; subtract it
      (let [err-poly (p/base-n-digits error 8)]
        (extract-data
         (p/- data err-poly GF8))))
    ))

(def RS-7-5-decode
  (partial RS-7-decode RS-7-5 RS-7-5-table))

(def RS-7-3-decode
  (partial RS-7-decode RS-7-3 RS-7-3-table))

(def RS-15-11
  "(15,11) Reed-Solomon code over GF(16)
  with generating polynomial
  g(x) = (x-w)(x-w^2)(x-w^3)(x-w^4), where w=2"
  (reduce
   (fn [p1 p2] (p/* p1 p2 GF16))
   [[2 1] [4 1] [8 1] [3 1]]))

(def BCH-15-7
  "(15,7) BCH binary code 2 errors.
  g(x) = (x^4+x+1)(x^4+x^3+x^2+x+1) is constructed so that
  w,w^2,w^3,w^4 are zeroes of g(x) in GF(16)"
  (p/* [1 1 0 0 1] [1 1 1 1 1] GF2))

(def BCH-15-5
  "(15,5) BCH binary code corrects 5 errors"
  (reduce
   (fn [p1 p2] (p/* p1 p2 GF2))
   [[1 1 0 0 1] [1 1 1 1 1] [1 1 1]]))

(def RS-255-223
  "(255,223) Reed-Solomon code over GF256 corrects 16 errors.
  g(x)=(x-w)(x-w^2)...(x-w^32)
  This is the standard recommended by CCSDS."
  (let [mul (:* GF256)
        exp (:exp GF256)
        prim 2
        roots (for [e (range 1 33)] (exp e))]
    (reduce
     (fn [p1 p2] (p/* p1 p2 GF256))
     (for [r roots] [r 1]))))

(defn syndromes
  "calculate n syndromes from the data polynomial"
  [poly n field]
  (let [mul (:* field)
        roots (:exp field)
        prim (:primitive field)
        roots (take n roots)]
    (vec (for [r roots]
           (p/evaluate poly r field)))))

(defn syndrome-mat
  [syns]
  (let [n (count syns)
        t (quot n 2)
        mat (make-array Integer/TYPE t t)]
    (doall (for [i (range t)
                 j (range t)]
             (aset mat i j
                   (int (syns (+ i j))))))
    mat))

(defn locate-n-errors
  "locate error positions for exactly n errors"
  [syns poly n field]
  (let [prim (:primitive field)
        mul (:* field)
        n-elts (count (:inv field))
        syns (vec (take (* 2 n) syns))
        syn-mat (syndrome-mat syns)
        syn-mat-inv (try
                      (la/mat-inv syn-mat field)
                      (catch Exception e nil))
        syn-vec (into-array Integer/TYPE
                            (subvec syns n (* 2 n)))]
    (if (nil? syn-mat-inv) nil          ; no errors to correct
        (let [err-vec (la/mat-vec syn-mat-inv syn-vec field)
              locator (reverse (conj err-vec 1))

              roots (:exp field)
              eval-roots (vec (map
                               #(p/evaluate locator % field)
                               roots))
              ;; zeros have the form w^k. this finds each exponent k
              locator-zeros-log (filter #(= 0 (eval-roots %))
                                         (range n-elts))]
          (vec (sort
                 ;; the locator polynomial zeros are inverses
                 ;; of the error locations
                 (map #(mod (- n-elts %) n-elts)
                      locator-zeros-log)))))
    ))

(defn error-sizes
  [syns idxs field]
  (let [t (count idxs)
        roots (:exp field)
        n (count roots)
        mat (make-array Integer/TYPE t t)]
    (do
      ;; solve linear equation to find error sizes from syndromes
      (doall (for [i (range t)
                   j (range t)]
               (aset mat i j
                     (int
                      (roots
                       (mod (* (inc i) (nth idxs j))
                            n))))
               ))
      (let [inv (try
                  (la/mat-inv mat field)
                  (catch Exception e nil))]
        (if (nil? inv) nil
            (let [syn-vec (into-array Integer/TYPE (take t syns))
                  size-vec (la/mat-vec inv syn-vec field)]
              size-vec)
            )))
    ))

(defn locate-errors
  "locate error positions for <= n errors"
  [syns poly nerrs field]
  (loop [t nerrs]
    (if (< nerrs 1) nil
        (let [err-idxs (locate-n-errors syns poly t field)]
          (if (nil? err-idxs)
            (recur (dec t))
            err-idxs)))
    ))

(defn decode
  "Peterson decoding algorithm for BCH or RS code"
  [poly nerrs field]
  (let [syns (syndromes poly (* 2 nerrs) field)]

    (cond (every? zero? syns) poly      ; no errors
          :else
          (let [err-idxs (locate-errors syns poly nerrs field)]
            (cond (or (nil? err-idxs) (empty? err-idxs))
                  nil
                  :else
                  {:locations err-idxs
                   :sizes (error-sizes syns err-idxs field)})))
    ))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; examples and junk

(comment
  (-> [5 2 3 1 6]
      (encode RS-7-5 GF8)
      (RS-7-5-decode))
  ;; => [5 2 3 1 6]
  (encode [1 6 3] RS-7-3 GF8)
  ;; => [0 2 2 4 1 6 3]
  (RS-7-3-decode [0 2 2 4 1 6 3])

  (RS-7-3-decode [1 2 2 4 1 7 3])
  ;; => [1 6 3]
  (RS-7-3-decode [0 2 2 4 6 6 6])
  ;; => [1 6 3]

  (encode [0 1 0 1 0 1 0] BCH-15-7 GF2)
  ;; => [0 1 0 1 1 0 0 0 0 1 0 1 0 1 0]

  ;; check 2 is a primitive element
  (let [mul (:* GF256)]
    (= (range 1 256)
       (sort (take 255 (iterate #(mul 2 %) 2))))))

(comment
  (def my-encoded-msg
    (let [message
          (str
           "Hello world! This is meiji163 transmitting from Neptune. "
           "It's cold here, Please send hot chocolate. Thanks. "
           "Now that I think of it, ramen would be good too if you have some.")

          data (vec (map int message))
          padded (p/shift-right
                  data
                  (- 223 (count data)))]
      (encode padded RS-255-223 GF256)))

  (let [err [1 42 1]
        max-errs 5
        decode-me (p/+ my-encoded-msg err GF256)
        syns (syndromes decode-me (* 2 max-errs) GF256)]
    (locate-errors syns decode-me max-errs GF256) ;; => [0 1 2]
    (decode decode-me max-errs GF256) ;; => {:locations [0 1 2], :sizes [1 42 1]}
    )

  (let [err [0 42 1 0 0 163 0 0 66 0 0 0 0 0 101 100]
        max-errs 8
        decode-me (p/+ my-encoded-msg err GF256)]
    (decode decode-me max-errs GF256))
  ;; => {:locations [1 2 5 8 14 15], :sizes [42 1 163 66 101 100]}
  )


src/ecc_clj/linalg.clj
(ns ecc-clj.linalg
  (:require [ecc-clj.poly :as p]))

(def not-invertible (new Exception "matrix is not invertible"))

(defn map-row!
  "my functional programming sin"
  [f row]
  (loop [i 0]
    (when (< i (count row))
      (aset row i
            (f (aget row i)))
      (recur (inc i)))))

(defn add-row! [r1 r2 plus]
  (loop [i 0]
    (when (< i (count r1))
      (aset r1 i
            (plus (aget r1 i) (aget r2 i)))
      (recur (inc i)))))

(defn swap-rows! [arr i j]
  (let [ai (aget arr i)
        aj (aget arr j)]
    (do
      (aset arr i aj)
      (aset arr j ai))))

(defn gauss-elimination
  [mat field]
  (let [{plus :+
         mul :*
         div :/
         sub :-
         zero :zero} field

        rows (count mat)
        cols (count (first mat))]
    (loop [i 0
           j 0]
      (cond
        (or (>= i rows) (>= j cols))
        mat

        (= zero (aget mat i j)) ;; find a pivot row and swap
        (let [swap-ix (first (filter
                              #(not= zero (aget mat % j))
                              (range rows)))]
          (when (nil? swap-ix) (throw not-invertible))

          (swap-rows! mat swap-ix i)
          (recur i j))

        :else
        ;; subtract current row from other rows
        ;; such that current column has all zeros
        (do
          (doall
           (map
            (fn [k] (when (not= i k)
                      (let [aij (aget mat i j)
                            akj (aget mat k j)
                            ai (aget mat i)
                            factor (if (= zero akj)
                                     0
                                     (div akj aij))
                            sub-row (amap ai i ret
                                          (mul factor (aget ai i)))]
                        (add-row! (aget mat k) sub-row plus))))
            (range rows)))
          ;; make this diagonal = 1
          (let [aij (aget mat i j)
                ai (aget mat i)]
            (map-row! #(div % aij) ai))
          ;;
          (recur (inc i) (inc j)))
        ))
    ))

(defn mat-inv [mat field]
  (let [rows (count mat)
        cols (count (first mat))]
    (if (not= rows cols)
      (throw (new Exception "matrix is not square"))

      ;; extend matrix with the identity matrix
      (let [exmat (make-array Integer/TYPE rows (* 2 cols))]
        (do
          (doall (for [i (range rows)
                       j (range cols)]
                   (aset exmat i j
                         (aget mat i j))))
          (doall (for [i (range rows)]
                   (aset exmat i (+ i rows)
                         (int 1))))

          ;; try gaussian elimination
          (let [elim (gauss-elimination exmat field)
                ;; it succeeded if the left half equals identity matrix
                success? (every? (fn [[i j]]
                                   (= (if (= i j) (int 1) (int 0))
                                      (aget elim i j)))
                                 (for [i (range rows)
                                       j (range cols)]
                                   [i j]))]

            (if (not success?) nil

                ;; extract the inverse matrix from right half
                (let [inv (make-array Integer/TYPE rows rows)]
                  (doall (for [i (range rows)
                               j (range cols)]
                           (aset inv i j
                                 (aget elim i (+ j rows)))))
                  inv))))
        ))
    ))

(defn dot [v1 v2 field]
  (let [{plus :+
         mul :*} field]
    (areduce v1 i res 0
             (plus res
                   (mul (aget v1 i)
                        (aget v2 i))))
    ))

(defn mat-vec
  "multiply matrix by vector.
  arr is an array representing a column vector."
  [mat arr field]
  (let [n (count arr)]
    (vec (for [i (range n)]
       (dot (aget mat i) arr field)))))

(defn matrix* [m1 m2 field]
  (let [{plus :+
         mul :*} field
        rows1 (count m1)
        cols1 (count (first m1))
        rows2 (count m2)
        cols2 (count (first m2))
        rows3 (count m1)
        res (make-array Integer/TYPE rows1 cols2)]
    (cond
      (or (not= cols1 rows2)) (throw (new Exception "incorrect matrix sizes"))
      :else
      "TODO")
    ))


src/ecc_clj/qr.clj
(ns ecc-clj.qr
  (:import java.awt.image.BufferedImage
           java.awt.Color
           javax.imageio.ImageIO
           java.io.File)
  (:require [ecc-clj.poly :as p]
            [ecc-clj.core :as ecc]))

(def GF256
  "GF(255) constructed as GF2[x]/<x^8+x^4+x^3+x^2+1>"
  (let [GF2-poly (p/parse-bin "100011101")]
    (p/char2-field 2 GF2-poly)))

;; ISO/IEC 18004 Annex A:
;; The (27,19) code is shortened from (255,248) code with this polynomial
;; w^21 + w^102 x + w^238 x^2 + w^149 x^3 + w^146 x^4 + w^229 x^5 + w^87 x^6 + x^7
(def RS-255-248
  "(255,248) Reed-Solomon code over GF256.
  g(x)=(x-w^0)(x-w^1)...(x-w^6)"
  (let [roots (take 7 (:exp GF256))]
    (reduce
     (fn [p1 p2] (p/* p1 p2 GF256))
     (for [r roots] [r 1]))))

;; https://en.wikipedia.org/wiki/QR_code#/media/File:QR_Character_Placement.svg
;;                         1                   2
;;     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0
;;
;;  0  1 1 1 1 1 1 1 0 f         0 1 1 1 1 1 1 1
;;  1  1 0 0 0 0 0 1 0 f         0 1 0 0 0 0 0 1
;;  2  1 0 1 1 1 0 1 0 f         0 1 0 1 1 1 0 1
;;  3  1 0 1 1 1 0 1 0 f         0 1 0 1 1 1 0 1
;;  4  1 0 1 1 1 0 1 0 f         0 1 0 1 1 1 0 1
;;  5  1 0 0 0 0 0 1 0 f         0 1 0 0 0 0 0 1
;;  6  1 1 1 1 1 1 1 0 1 0 1 0 1 0 1 1 1 1 1 1 1
;;  7  0 0 0 0 0 0 0 0 f         0 0 0 0 0 0 0 0
;;  8  f f f f f f 1 f f         f f f f f f f f
;;  9              0                           x
;; 10              1                           x
;; 11              0                           x
;; 12              1                           x
;; 13  0 0 0 0 0 0 0 0 1                       x
;; 14  1 1 1 1 1 1 1 0 f                       x
;; 15  1 0 0 0 0 0 1 0 f                       x
;; 16  1 0 1 1 1 0 1 0 f                       x
;; 17  1 0 1 1 1 0 1 0 f                       x
;; 18  1 0 1 1 1 0 1 0 f                       x
;; 19  1 0 0 0 0 0 1 0 f                       x
;; 20  1 1 1 1 1 1 1 0 f                       x

(def fixed1s
  (let [sqr [[0 0] [0 1] [0 2] [0 3] [0 4] [0 5] [0 6]
             [6 0] [6 1] [6 2] [6 3] [6 4] [6 5] [6 6]
             [1 0] [2 0] [3 0] [4 0] [5 0]
             [1 6] [2 6] [3 6] [4 6] [5 6]
             ;;
             [2 2] [2 3] [2 4]
             [3 2] [3 3] [3 4]
             [4 2] [4 3] [4 4]]]
    (vec (concat
          sqr
          (for [[i j] sqr] [(+ i 14) j])
          (for [[i j] sqr] [i (+ j 14)])
          [[13 8] ;; one unused format bit
           [8 6] [10 6] [12 6]
           [6 8] [6 10] [6 12]]))))

(def fixed0s
  (let [sqr [[1 1] [1 2] [1 3] [1 4] [1 5]
             [5 1] [5 2] [5 3] [5 4] [5 5]
             [2 1] [3 1] [4 1]
             [2 5] [3 5] [4 5]
             ;;
             [0 7] [1 7] [2 7] [3 7] [4 7] [5 7] [6 7] [7 7]
             [7 0] [7 1] [7 2] [7 3] [7 4] [7 5] [7 6]]]
    (vec (concat
          sqr
          (for [[i j] sqr] [(- 20 i) j])
          (for [[i j] sqr] [i (- 20 j)])
          [[9 6] [11 6]
           [6 9] [6 11]]))))

(defn v-pattern
  "(len x 2) vertical blocks either :up or :down"
  [dir xy len]
  (let [[row col] xy
        rng (cond
              (= dir :up) (reverse
                           (range (inc (- row len)) (inc row)))
              (= dir :down) (range row (+ row len))
              :else
              (throw (IllegalArgumentException.)))]
    (->> (for [r rng] [[r col] [r (dec col)]])
         (reduce concat)
         (vec))))

(defn h-pattern
  "(2 x 8) horizonal blocks either :cwise (clockwise) or :anticwise"
  [dir xy]
  (let [[row col] xy
        r+1 (inc row) r-1 (dec row)
        c-1 (- col 1) c-2 (- col 2) c-3 (- col 3)]
    (cond (= dir :cwise)
          ;; starts at upper left corner
          [[row col] [row c-1] [r+1 col] [r+1 c-1]
           [r+1 c-2] [r+1 c-3] [row c-2] [row c-3]]

          (= dir :anticwise)
          ;; starts at lower left corner
          [[row col] [row c-1] [r-1 col] [r-1 c-1]
           [r-1 c-2] [r-1 c-3] [row c-2] [row c-3]]
          :else
          (throw (IllegalArgumentException.)))))

(def v1-bit-order
  "2d coords of data bits in the order they are filled"
  (vec (reduce concat
    [(v-pattern :up [20 20] 10) (h-pattern :anticwise [10 20])
     (v-pattern :down [11 18] 8) (h-pattern :cwise [19 18])
     (v-pattern :up [18 16] 8) (h-pattern :anticwise [10 16])
     (v-pattern :down [11 14] 8) (h-pattern :cwise [19 14])
     (v-pattern :up [18 12] 12)
     ;;
     (v-pattern :up [5 12] 4) (h-pattern :anticwise [1 12])
     (v-pattern :down [2 10] 4)
     (v-pattern :down [7 10] 14)
     ;;
     (v-pattern :up [12 8] 4) (v-pattern :down [9 5] 4)
     (v-pattern :up [12 3] 4) (v-pattern :down [9 1] 4)])))

(def v1-fmt-order
  "coords of format bits. 2 copies of (5 data bits + 10 check bits)"
  [[8 0] [8 1] [8 2] [8 3] [8 4]
   [8 5] [8 7] [8 8] [7 8] [5 8] [4 8] [3 8] [2 8] [1 8] [0 8]
   ;;
   [20 8] [19 8] [18 8] [17 8] [16 8]
   [15 8] [14 8] [8 13] [8 14] [8 15] [8 16] [8 17] [8 18] [8 19] [8 20]])

(def fmt-bitmask
  [1 0 1 0 1 0 0 0 0 0 1 0 0 1 0])

(defn fmt-bits
  [mask]
  (let [mb [0 0 1]
        ec-level [0 1]
        bits (reverse (concat ec-level mb))
        enc-bt (ecc/encode bits ecc/BCH-15-5 ecc/GF2)]
    (p/+ (reverse enc-bt) fmt-bitmask ecc/GF2)))

(def alphanum
  (let [chrs "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ $%*+-./:"]
    (zipmap chrs (range (count chrs)))))

(defn encode-bytes
  "encode 17 bytes to bit vector in the format specified by QR v1"
  [bs]
  (if (not= 17 (count bs)) (throw (IllegalArgumentException.))
      (let [tobits #(p/bits-to-vec % {:pad 8 :order :big})

            len (tobits (count bs))
            end [0 0 0 0]
            enc [0 1 0 0]
            data-bits (reduce concat (map tobits bs))
            msg-bits  (reduce concat [enc len data-bits end])

            ;; NB: have to reverse twice!
            data-poly (->> msg-bits
                           (p/bits-to-bytes)
                           (reverse))
            check-poly (p/mod (p/shift-right data-poly 7)
                              RS-255-248 GF256)
            check-bits (->> check-poly
                            (reverse)
                            (map tobits)
                            (reduce concat))]
        (concat msg-bits check-bits))))

(def mask-preds
  [(fn [i j] (even? (+ i j)))
   (fn [i j] (even? i))
   (fn [i j] (= 0 (mod (+ i j) 3)))
   (fn [i j] (= 0 (mod j 3)))
   (fn [i j] (even? (+ (quot j 3) (quot i 2))))
   (fn [i j] (= 0 (+ (mod (* i j) 2) (mod (* i j) 3))))
   (fn [i j] (even? (+ (* i j) (mod (* i j) 3))))
   (fn [i j] (even? (+ i j (mod (* i j) 3))))])

(defn mask-bits
  "apply one of the QR bitmasks on the bit sequence. The mask codes are 0-7."
  [code bits]
  (let [mask? (mask-preds code)
        coord-bits (map vector v1-bit-order bits)]
    (for [[[i j] bit] coord-bits]
      (if (mask? i j)
        (bit-xor 1 bit) bit))))

(defn fill-sqr [g size coord color]
  (let [[i j] coord]
    (.setColor g color)
    (.fill g (java.awt.Rectangle. i j size size))))

(defn qr-image
  "create 21x21 QR code in BufferedImage given the format bits"
  [fmtbits databits & [scale]]
  (if (or (not= 15 (count fmtbits))
          (not= (* 26 8) (count databits)))
    (throw (IllegalArgumentException.))

    (let [scale (or scale 10)
          size (* scale (+ 21 2))
          white Color/WHITE
          black Color/BLACK
          img (BufferedImage. size size BufferedImage/TYPE_INT_RGB)
          g (.createGraphics img)

          data-coord-bits (map vector v1-bit-order databits)
          fmt-coord-bits (map vector v1-fmt-order
                              (concat fmtbits fmtbits))
          ;; coords on java image are swapped from QR coords
          img-coord (fn [[i j]]
                      [(* scale (inc j)) (* scale (inc i))])]

      (fill-sqr g size [0 0] white)
      (doall (for [crd fixed1s]
               (fill-sqr g scale (img-coord crd) black)))
      (doall (for [crd fixed0s]
               (fill-sqr g scale (img-coord crd) white)))
      (doall (for [[crd b] data-coord-bits]
               (fill-sqr g scale (img-coord crd)
                         (if (= 0 b) white black))))
      (doall (for [[crd b] fmt-coord-bits]
               (fill-sqr g scale (img-coord crd)
                         (if (= 0 b) white black))))
      (.dispose g)
      img)))

(comment
  (let [mask-code 1
        databits (->> "github.com/github"
                      (map int)
                      (encode-bytes)
                      (mask-bits mask-code))
        fmtbits (fmt-bits mask-code)
        myimg (qr-image fmtbits databits)]
    (ImageIO/write myimg "PNG" (File. "test.png")))
  )
